Bwoink System Refactor

Simyon264 opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
This PR refactors the entirety of Admin Help. 

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
Admin Help was an *old* **old** system not up to standards at all. This PR refactors admin help, bringing it to modern codebase standards. This PR is made as a draft PR to get feature requests in and to allow early reviews where possible so that I do not have to refactor a refactor.

Fixes #37034 Fixes #35523 (Possibly) Fixes #23593

You can track current progress on this PR here: 
https://jetfish.iterator.systems/projects/0199d82e-3c62-7e94-b61e-2c81db27118e

### Use this link to view all current capabilities of the system and planned features.

## Technical details
<!-- Summary of code changes for easier review. -->
The bulk of the refactor is to atomize admin help. Admin Help is now based on prototypes.
```yaml
- type: bwoinkChannel
  id: AdminHelp
  name: bwoink-channel-ahelp
  helpText: bwoink-channel-ahelp-help
  order: -1 # We always want this one to be the first
  manageRequirement:
    requirements:
    - !type:AdminFlagRequirement
      flags: Adminhelp
  features:
  - !type:SoundOnMessage
    sound: !type:SoundPathSpecifier
      path: /Audio/Animals/cat_meow.ogg
  - !type:ManagerOnlyMessages
    prefix: bwoink-message-admin-only
    checkName: admin-ahelp-admin-only
  - !type:DiscordRelay
    channelCvar: discord.bwoink.channel.ahelp
  - !type:StatusMessages

```

This allows modular adding of features to admin help without hardcoding many of the values.

The core pillar is that admin help is now split into "channels". These channels are shown via a selector in the F1 menu. The intent is that adding something like a "discord relay" is as simple as adding that to the feature of the channel and from there a separate manager goes and handles it. 

**For how this entire code is now structured:**
`BwoinkSystem` is now `BwoinkManager`.
`SharedBwoinkManager` handles the shared utility methods. `ServerBwoinkManager` and `ClientBwoinkManager` handle the relevant implementations on their respective sides.

The server and client both hold a `Conversations` Dictionary. It is synced to the client, allowing for admins to see the full history of a player. Systems and other managers can easily hook into the admin help system via `MessageReceived`.
A message object holds the properties as you would expect: Sender, Time, Content and **importantly the message flags**.
The old system had messages use a complicated case of "just add more properties to the message object". While this works I opted for the more extendable system of adding an Enum to the message object "MessageFlags":
```cs
[Flags]
public enum MessageFlags : byte
{
    None = 0,

    /// <summary>
    /// The person who sent this message is a manager for this channel.
    /// </summary>
    Manager = 1,

    /// <summary>
    /// This message should not result in a sound.
    /// </summary>
    Silent = 2,

    /// <summary>
    /// This message can only be seen by other managers of this channel.
    /// </summary>
    ManagerOnly = 4,

	// (...)
}
```

### Granular permissions:

Channels can be written, read and managed. 
**Write**: Write allows a person who cannot manage to send message to managers. 
**Read**: Read allows a non manager to receive messages from managers
**Manage**: Manage implicitly allows Write and Read but importantly allows write and read into user channels which aren't your own.

For checking access, the server and client both have the requirements API which calculate down to a `BwoinkChannelConditionFlags` enum. This API can be hooked into server side via `OnManageCheck` `OnReadCheck` and `OnWriteCheck`.
Since some stuff may be server only, the client only keeps a copy of the permissions that the server sends over. If you hook into any of the check events, ensure you are calling `SyncChannels` when appropriate.

These permissions all are based on `BwoinkChannelRequirement` which has the following fields:
**OperationMode**: This field determines how to evaluate the conditions contained. This must be either: All, Any, InvertedAny or None.
**Requirements**: This field contains the list of requirements that must be met (or not, depending on the mode)

Adding new conditions/requirements involves making a new Requirement file in which the class inherits the BwoinkChannelCondition class. Then adding the handler in either the `ServerBwoinkManager.Requirements.cs` file or a handler for the mentioned events above.

### Discord Relay

The entirety of the relay is contained within `BwoinkDiscordRelayManager`.
It listens for new messages via the `MessageReceived` event. And every 2 seconds queues an update (if one isn't on-going).

Messages themselves are formatted via the shared public API `SharedBwoinkManager.FormatMessage`. For supporting bold tags I do some shitcode which places `**` for each bold tag we encounter. 

### Status Messages

The status messages (X has disconnected, X was banned, etc.) are contained in `MessageBwoinkManager`. This manager is also responsible for implementing the (for now) global `AdminAhelpOverrideClientName` cvar. For implementing the custom emojis for the Discord relay I (kind of shitcode) hijacked the Sender of the message to contain info about the message type. This is an easy workaround since I didn't feel like making an entire custom payload system for messages. *If the sender is already not being shown, why not use it?*.

### Public API

To send message server side, you can use the 3 `SendMessageInChannel` methods. They do all the magic required for state sync. 
To modify a message before it is sent to a client, you can listen to the `MessageBeingSent` event on the server. This event is only fired for non-manager clients. *Importantly: This does not modify the message stored on the server*

### Considerations for the future

Since I removed the action buttons in favor of the player panel, the player panel **must be refactored after this PR and before this PR hits stable**. This is to make the admin experience not ass :tm:

**Pruning old messages:**
This is also a requirement for stable (possibly). Not done in this PR because I want to finish this shit and it is not that much of a performance drain.

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->
https://memory-conflux.iterator.systems/bwoink%20refactor_1.mp4
https://memory-conflux.iterator.systems/bwoink%20refactor_2.mp4
https://memory-conflux.iterator.systems/bwoink%20refactor_3.mp4 **(latest)**
<img width="549" height="401" alt="image" src="https://github.com/user-attachments/assets/612a7152-de63-4b56-9674-80a7f21ebdac" />
<img width="577" height="186" alt="image" src="https://github.com/user-attachments/assets/81082914-82ae-4743-b6f9-f5cda0bdd121" />


## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

`BwoinkSystem.cs` was removed on the client and server. Use `SharedBwoinkManager.cs`

For sending messages into AHelp channels, use `ServerBwoinkManager.SendMessageInChannel`. 

The following CVars were removed:
- discord.on_call_ping
- discord.on_call_webhook
- discord.ahelp_webhook
- discord.ahelp_footer_icon
- discord.ahelp_avatar
- audio.ahelp_sound

`DiscordLink.SendMessageAsync` now has an optional embeds argument and returns a nullable RestMessage

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
ADMIN:
:cl: Simyon
- add: AHelp channels can now be prototype uploaded.
- fix: AHelp history is now preserved across disconnects.
- add: Added bwoink:message, bwoink:addchannel, bwoink:rmchannel and bwoink:sync toolshed commands.
- fix: The AHelp Discord relay will now escape markdown
- add: The AHelp Discord relay will now show a next button.

Simyon264 committed:
Nuke BwoinkSystem.cs

Simyon264 committed:
WIP: Ahelp refactor

Simyon264 committed:
Remove leftover AHelpMessage

Simyon264 committed:
Add doc-comment to Flags

Simyon264 committed:
Make message flags a byte

Simyon264 committed:
Fix lobby button going out of sync if you close window manually

Simyon264 committed:
Clear text output when syncing lines

Simyon264 committed:
add channel help text when clearing once again

Simyon264 committed:
Remove unused method for now

Simyon264 committed:
implement message sync

Simyon264 committed:
fix admins sending in their own channel resulting in no messages being sent

Simyon264 committed:
don't send help text for manager panels

Simyon264 committed:
properly handle rich text

Simyon264 committed:
convert into method group

Simyon264 committed:
Make unreads work properly

Simyon264 committed:
minor, make props readonly

Simyon264 committed:
Fix some issues with UI lifetime (i think)

Simyon264 committed:
Fix unclosed menu not keeping unread count

Simyon264 committed:
also do this for last message

Simyon264 committed:
give nik the sound they deserve

Simyon264 committed:
ui manager isn't real

Simyon264 committed:
add support for channels without sound play

Simyon264 committed:
remove unused imports

Simyon264 committed:
Register a shared instance for bwoink manager

Simyon264 committed:
there IS a better way of doing this

Simyon264 committed:
it did, in fact, not work

Simyon264 committed:
i have, the stupid

Simyon264 committed:
i have the stupid disease

Simyon264 committed:
implement typing status

Simyon264 commented:
does it NOT do that currently???

Admiral-Obvious-001 commented:
> does it NOT do that currently???

Nope. Sometimes it just doesn't follow the player at all.

Optionally, if the player disconnected, it should follow their last body instead.

luckyshotpictures commented:
> does it NOT do that currently???

No, it does not. Whenever a user updates their mob (ghosts, polymorph, etc) you need to click off and on again for the entity to be correct.

Kowlin commented on Content.Client/Administration/UI/Bwoink/BwoinkControl.xaml:
Please keep the quick actions for player interactions. Like follow, kick, ban, etc. I'm afraid that this will ruin the experience that admins have come to expect with the ahelp menu, also the player panel is frankly extremely slow and is unusable till data from the database is loaded in.

sleepyyapril replied:
My opinion on the actions list is there's definitely a way better way to do it than hardcoded buttons lmao

Simyon264 committed:
Merge branch 'master' into bwoink-system-refactor

# Conflicts:
#	Content.Client/Administration/UI/Bwoink/BwoinkControl.xaml
#	Content.Client/UserInterface/Systems/Bwoink/AHelpUIController.cs
#	Content.Server/Entry/EntryPoint.cs
#	Content.Server/IoC/ServerContentIoC.cs

Simyon264 committed:
fix conflicts

Simyon264 committed:
cleanup and add more comments

Simyon264 commented:
As an FYI, this PR could in theory also include Mentor Help. Just requires game admin signoff on how they want it.

Simyon264 commented:
https://memory-conflux.iterator.systems/bwoink%20refactor_1.mp4

Prototype uploading and the separate channels demonstrated.

Simyon264 committed:
Add public API and toolshed command for bwoinking

Simyon264 committed:
properly switch to channel when getting bwoinked in that channel

Simyon264 committed:
add temporary testing channel

Simyon264 committed:
half working implementation of better requirements for channels

im so good at this

Simyon264 committed:
finish list requirement and handling channel syncs

Simyon264 committed:
Implement OperationMode.None

Allows to fully deny a requirement, always.

Simyon264 committed:
add sync command

Synchronizes the entire bwoink state for a given session. This can be done in cases where the Conversations change server side (for example VV) and the client is not told about it.

Simyon264 committed:
unsub from cvar value

Simyon264 committed:
remove unused import

Simyon264 committed:
admin only messages

Simyon264 committed:
add proper access checks to message command

Simyon264 committed:
Sync channels on prototype reload

Simyon264 committed:
make InvokeMessageReceived less stinky

Simyon264 committed:
implement silent messages

Simyon264 committed:
Add ratelimits

Simyon264 committed:
make playerlist a bit less wide by default

Simyon264 committed:
make tab title show amount of unread messages

Simyon264 committed:
readd relay label

Simyon264 committed:
Rename to ServerBwoinkManager

Simyon264 committed:
Move FormatMessage into shared

Simyon264 committed:
setup infra for discord relay

Simyon264 committed:
Merge remote-tracking branch 'origin/master' into bwoink-system-refactor

# Conflicts:
#	Content.Server/Administration/Systems/BwoinkSystem.cs

Simyon264 committed:
fix note test

why was it even using the ahelp menu ????

Simyon264 committed:
fix ratelimits

I put the target player as the rate limit user, whoops.

Simyon264 committed:
implement NoReceivers flag

Simyon264 committed:
see previous

Simyon264 committed:
add very basic discord relay

Simyon264 committed:
fix minor spelling mistake

life ruined

Simyon264 committed:
fix senderid being set to null serverside when filtering

Simyon264 committed:
fix window title not being set

Simyon264 committed:
add events for requirement checks

Simyon264 committed:
add MessageBeingSent event on ServerBwoinkManager
add support for AdminAhelpOverrideClientName cvar

:3

Simyon264 committed:
make deps readonly

Simyon264 committed:
make MessageBeingSent an event

Simyon264 committed:
add GetBwoinkChannelsWithFeature API

Simyon264 committed:
make SynchronizeMessage work with non-connected clients

Simyon264 committed:
don't send "cant read" message when target is offline

Simyon264 committed:
Add system message type

Simyon264 committed:
Add SendMessageInChannel overload that takes a BwoinkMessage object

we love overloads

Simyon264 committed:
Add status messages

Simyon264 committed:
make relay messages escape markdown and make sender bold

Simyon264 committed:
make non-manager ahelp not be giant as fuck

Simyon264 committed:
fix minor ui skill issues

Simyon264 committed:
Finish todos

Simyon264 committed:
the final bwoink....

Simyon264 committed:
add myself as codeowner

Princess-Cheeseballs reviewed (changes requested):
We need the buttons back. This is non-negotiable. When dealing with a raider seconds matter and having to navigate to the playerpanel and wait for it to load is not acceptable. Being unable to quickly grab notes when processing an ahelp is not acceptable. 

This is a massive disruption to admin workflow and they need to be added back in this PR.

Simyon264 committed:
don't queue messages into relay when link not connected

Simyon264 committed:
add integration tests

Simyon264 committed:
remove unneeded UpdatePlayerRecords call

Simyon264 committed:
i hate it here

Simyon264 committed:
a

Simyon264 committed:
add custom action buttons

Simyon264 committed:
clean up tests some more

Simyon264 committed:
mayhaps fix tests?

PJB3005 commented on Resources/Prototypes/bwoink_channels.yml:
Almost all of this can and should just be defined via C# logic instead.

Simyon264 replied:
The idea is that admins can prototype upload channels for use (for example) in events as a help channel.
That way the system is expandable and flexible.

PJB3005 replied:
Yeah, no, we're not supporting that.
